File: top_interview_quest/easy/romantoint.py
class Solution:
    def romanToInt(self, s: str) -> int:
        plan = {'I':1, 'V':5,'X':10,'L':50,'C':100, 'D':500, 'M':1000}
        num = 0
        for i,j in enumerate(s):
                if i<(len(s)-1):
                    if (j=='I') and (s[i+1]=='V' or s[i+1]=='X'):
                        num-=plan[j]
                        continue
                    elif (j=='X') and (s[i+1]=='L' or s[i+1]=='C'):
                        num-=plan[j]
                        continue
                    elif (j=='C') and (s[i+1]=='D' or s[i+1]=='M'):
                        num-=plan[j]
                        continue
                num+=plan[j]
        return num   




##alternative method

class Solution:
    def romanToInt(self, s: str) -> int:
        translations = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
        }
        number = 0
        s = s.replace("IV", "IIII").replace("IX", "VIIII")
        s = s.replace("XL", "XXXX").replace("XC", "LXXXX")
        s = s.replace("CD", "CCCC").replace("CM", "DCCCC")
        for char in s:
            number += translations[char]
        return number

File: top_interview_quest/easy/romantoint.cpp
#include <vector>
class Solution {
public:
    int romanToInt(string s) {
        int num = 0;
        vector<char> roman{'I','V','X','L','C','D','M'};
        vector<int> number{1,5,10,50,100,500, 1000};
        for (int i=0;i<s.size();i++) {
            auto it = find(roman.begin(),roman.end(),s[i]);
            if (it!=roman.end()) {
                int ind = int(it - roman.begin());
                if (i<s.size()-1) {
                    if ((s[i]=='I') && ((s[i+1]=='V') || (s[i+1]=='X'))) {
                        num-=number[ind];
                        continue;
                    }
                    else if ((s[i]=='X') && ((s[i+1]=='L') || (s[i+1]=='C'))) {
                        num-=number[ind];
                        continue;
                    }
                    else if ((s[i]=='C') && ((s[i+1]=='D') || (s[i+1]=='M'))){
                        num-=number[ind];
                        continue;
                    }
                }         
             num+=number[ind]; }
        }
        
        return num;
        
    }
};



//alternative
int romanToInt(string s) 
{
    unordered_map<char, int> T = { { 'I' , 1 },
                                   { 'V' , 5 },
                                   { 'X' , 10 },
                                   { 'L' , 50 },
                                   { 'C' , 100 },
                                   { 'D' , 500 },
                                   { 'M' , 1000 } };
                                   
   int sum = T[s.back()];
   for (int i = s.length() - 2; i >= 0; --i) 
   {
       if (T[s[i]] < T[s[i + 1]])
       {
           sum -= T[s[i]];
       }
       else
       {
           sum += T[s[i]];
       }
   }
   
   return sum;
}


